File: internal/adapter/adapter.go
package adapter

import (
	"fmt"
	"sync"

	"github.com/justlovemaki/AIClient-2-API/internal/common"
)

// ApiServiceAdapter defines the interface for all API service adapters
type ApiServiceAdapter interface {
	// GenerateContent generates content using the API
	GenerateContent(model string, requestBody map[string]interface{}) (map[string]interface{}, error)

	// GenerateContentStream generates content using the API in streaming mode
	GenerateContentStream(model string, requestBody map[string]interface{}) (<-chan interface{}, error)

	// ListModels lists available models
	ListModels() (map[string]interface{}, error)

	// RefreshToken refreshes the authentication token if needed
	RefreshToken() error

	// IsInitialized returns whether the adapter is initialized
	IsInitialized() bool
}

var (
	adapterInstances = make(map[string]ApiServiceAdapter)
	adapterMutex     sync.RWMutex
)

// GetAdapter returns an adapter instance for the given configuration
func GetAdapter(config *common.Config) (ApiServiceAdapter, error) {
	provider := config.ModelProvider
	providerKey := provider
	if config.UUID != "" {
		providerKey = provider + config.UUID
	}

	adapterMutex.RLock()
	if adapter, exists := adapterInstances[providerKey]; exists {
		adapterMutex.RUnlock()
		return adapter, nil
	}
	adapterMutex.RUnlock()

	// Create new adapter
	adapterMutex.Lock()
	defer adapterMutex.Unlock()

	// Double-check after acquiring write lock
	if adapter, exists := adapterInstances[providerKey]; exists {
		return adapter, nil
	}

	var adapter ApiServiceAdapter
	var err error

	switch provider {
	case common.ModelProviderGeminiCLI:
		adapter, err = NewGeminiAdapter(config)
	case common.ModelProviderOpenAICustom:
		adapter, err = NewOpenAIAdapter(config)
	case common.ModelProviderClaudeCustom:
		adapter, err = NewClaudeAdapter(config)
	case common.ModelProviderKiroAPI:
		adapter, err = NewKiroAdapter(config)
	case common.ModelProviderQwenAPI:
		adapter, err = NewQwenAdapter(config)
	default:
		return nil, fmt.Errorf("unsupported model provider: %s", provider)
	}

	if err != nil {
		return nil, fmt.Errorf("failed to create adapter for %s: %w", provider, err)
	}

	adapterInstances[providerKey] = adapter
	fmt.Printf("[Adapter] Created adapter for %s (key: %s)\n", provider, providerKey)

	return adapter, nil
}

// GetAllAdapters returns all initialized adapter instances
func GetAllAdapters() map[string]ApiServiceAdapter {
	adapterMutex.RLock()
	defer adapterMutex.RUnlock()

	result := make(map[string]ApiServiceAdapter, len(adapterInstances))
	for k, v := range adapterInstances {
		result[k] = v
	}

	return result
}

// InitializeAdapters initializes adapters for all configured providers
func InitializeAdapters(config *common.Config) error {
	var errors []error

	// Initialize adapters for all default providers
	for _, provider := range config.DefaultModelProviders {
		providerConfig := &common.Config{
			ModelProvider:             provider,
			RequiredAPIKey:            config.RequiredAPIKey,
			ServerPort:                config.ServerPort,
			Host:                      config.Host,
			OpenAIAPIKey:              config.OpenAIAPIKey,
			OpenAIBaseURL:             config.OpenAIBaseURL,
			ClaudeAPIKey:              config.ClaudeAPIKey,
			ClaudeBaseURL:             config.ClaudeBaseURL,
			GeminiOAuthCredsBase64:    config.GeminiOAuthCredsBase64,
			GeminiOAuthCredsFilePath:  config.GeminiOAuthCredsFilePath,
			KiroOAuthCredsBase64:      config.KiroOAuthCredsBase64,
			KiroOAuthCredsFilePath:    config.KiroOAuthCredsFilePath,
			QwenOAuthCredsFilePath:    config.QwenOAuthCredsFilePath,
			ProjectID:                 config.ProjectID,
			SystemPromptFilePath:      config.SystemPromptFilePath,
			SystemPromptMode:          config.SystemPromptMode,
			SystemPromptContent:       config.SystemPromptContent,
			PromptLogBaseName:         config.PromptLogBaseName,
			PromptLogMode:             config.PromptLogMode,
			PromptLogFilename:         config.PromptLogFilename,
			RequestMaxRetries:         config.RequestMaxRetries,
			RequestBaseDelay:          config.RequestBaseDelay,
			CronNearMinutes:           config.CronNearMinutes,
			CronRefreshToken:          config.CronRefreshToken,
			DefaultModelProviders:     config.DefaultModelProviders,
		}

		fmt.Printf("[Initialization] Initializing adapter for %s...\n", provider)
		if _, err := GetAdapter(providerConfig); err != nil {
			fmt.Printf("[Initialization Warning] Failed to initialize adapter for %s: %v\n", provider, err)
			errors = append(errors, err)
		}
	}

	if len(errors) > 0 {
		return fmt.Errorf("failed to initialize %d adapter(s)", len(errors))
	}

	return nil
}

File: rust/src/adapter.rs
/*!
 * API Service Adapter
 *
 * Defines the adapter interface and implementations for different AI service providers.
 */

use crate::common::*;
use anyhow::Result;
use async_trait::async_trait;
use futures::Stream;
use std::pin::Pin;

/// Trait defining the interface for all AI service adapters
#[async_trait]
pub trait ApiServiceAdapter: Send + Sync {
    /// Generate content (non-streaming)
    async fn generate_content(
        &self,
        model: &str,
        request_body: serde_json::Value,
    ) -> Result<serde_json::Value>;

    /// Generate content (streaming)
    async fn generate_content_stream(
        &self,
        model: &str,
        request_body: serde_json::Value,
    ) -> Result<Pin<Box<dyn Stream<Item = Result<serde_json::Value>> + Send>>>;

    /// List available models
    async fn list_models(&self) -> Result<ModelListResponse>;

    /// Refresh authentication token (if applicable)
    async fn refresh_token(&self) -> Result<()>;
}

/// Factory function to create appropriate adapter based on provider type
pub async fn create_adapter(
    provider: ModelProvider,
    config: &crate::config::Config,
) -> Result<Box<dyn ApiServiceAdapter>> {
    match provider {
        ModelProvider::GeminiCliOAuth => {
            let service = crate::providers::gemini::GeminiApiService::new(
                config.gemini_oauth_creds_base64.clone(),
                config.gemini_oauth_creds_file_path.clone(),
                config.project_id.clone(),
                config.request_max_retries,
                config.request_base_delay,
            ).await?;
            Ok(Box::new(service))
        }
        ModelProvider::OpenAICustom => {
            let api_key = config.openai_api_key.clone()
                .ok_or_else(|| anyhow::anyhow!("OpenAI API key is required"))?;
            let service = crate::providers::openai::OpenAIApiService::new(
                api_key,
                config.openai_base_url.clone(),
                config.request_max_retries,
                config.request_base_delay,
            )?;
            Ok(Box::new(service))
        }
        ModelProvider::ClaudeCustom => {
            let api_key = config.claude_api_key.clone()
                .ok_or_else(|| anyhow::anyhow!("Claude API key is required"))?;
            let service = crate::providers::claude::ClaudeApiService::new(
                api_key,
                config.claude_base_url.clone(),
                config.request_max_retries,
                config.request_base_delay,
            )?;
            Ok(Box::new(service))
        }
        ModelProvider::ClaudeKiroOAuth => {
            let service = crate::providers::kiro::KiroApiService::new(
                config.kiro_oauth_creds_base64.clone(),
                config.kiro_oauth_creds_file_path.clone(),
                config.request_max_retries,
                config.request_base_delay,
            ).await?;
            Ok(Box::new(service))
        }
        ModelProvider::OpenAIQwenOAuth => {
            let service = crate::providers::qwen::QwenApiService::new(
                config.qwen_oauth_creds_file_path.clone(),
                config.request_max_retries,
                config.request_base_delay,
            ).await?;
            Ok(Box::new(service))
        }
    }
}

